### PR TITLE
Add two new number theory functions to the Integer trait

### PR DESCRIPTION
**Pull request by ejmahler**
Made as a part of issue #202 

This PR adds two methods to the Integer trait. The first is modular exponentiation (`pow_mod`). The second is multiplicative inverse modulo (`inverse_mod`).

I'd appreciate some scrutiny on my big int implementations of these functions. Although they pass all the tests, I'm sure there are optimizations to be made to avoid unnecessarily making copies of numbers, etc.

### CODE REVIEW

File: bigint/src/lib.rs [hunk 1]
```diff
@@ -1263,6 +1263,65 @@ impl Integer for BigUint {
...
+            One::one()
+        } else {
+            base = base % modulo;
+
+            while !exp.is_zero() {
+                if exp.is_odd() {
+                    result = result * &base % modulo;
+                }
+                exp = exp >> One::one();
+                base = &base * &base % modulo;
```
Comment (vks) [inline bigint/src/lib.rs]: You probably could avoid a copy by taking one `base` by value.

File: bigint/src/lib.rs [hunk 2]
```diff
@@ -2401,6 +2460,79 @@ impl Integer for BigInt {
...
```
Comment (vks) [inline bigint/src/lib.rs]: One `base` could be taken by value.

**Comment by vks** [comment]
There seems to be quite a bit of code duplication with regards to Euclid's algorithm to calculate the GCD. It would be nice to have that as a separate function, because it can be useful on its own. Also it would make it easier to replace with a more efficient binary algorithm.

**Comment by ejmahler** [comment]
Unfortunately, the GCD calculation is inseparable from the multiplicative inverse calculation. We're not merely calculating the GCD in order to verify that an inverse exists, the inverse calculation is an "addition" to the euclidean algorithm: https://en.wikipedia.org/wiki/Extended_Euclidean_algorithm

**Comment by vks** [comment]
You can implement the extended Euclidean algorithm as a separate function. See [my take](https://github.com/vks/discrete-log/blob/master/src/main.rs) on implementing the modular power and inverse.

**Comment by ejmahler** [comment]
Definitely - I decided not to so that we aren't doing the extra work of computing the other coefficient. If you think it's a better idea to compute it and ignore it in the name of cleaner code I'd be happy to do that.

If we go that route, how should we handle unsigned numbers with the extended euclidean algorithm? It commonly returns negative numbers. Should we just not implement it for unsigned numbers?

**Comment by vks** [comment]
You could also have a function that does not calculate the second coefficient. If you want to go full DRY, you could use a macro and dead code elimination.

I think for unsigned numbers Euclid's extended algorithm only makes sense if you implent it for modular numbers only.

**Comment by ejmahler** [comment]
If we're not implementing it for unsigned types, then it obviously can't be a method on the Integer trait. Where do you think it should go instead?

**Comment by cuviper** [comment]
I would definitely like to include unsigned types for `pow_mod` and `inverse_mod`, even if that means it can't use a "pure" version of extended GCD.

**Comment by ejmahler** [comment]
Closing because I'm getting the vibe this isn't acceptable without pretty big changes

**Pull request closed without merging**